packages/SwingSet/src/vats/comms/clist-kernel.js

```diff
@@ -1,145 +1,145 @@
-        // to another, we won't have mentioned it to the kernel yet. Rather
-        // than keep track of this, we just use a fresh vpid each time we
-        // need to talk about a resolved promise to the kernel. We must send
-        // the resolution and then immediately retire the vpid again.
-
-        const fresh = allocateResolvedPromiseID();
-        // console.log(`fresh: ${fresh} for ${vpid}`, p.rejected, p.data);
-        // we must tell the kernel about the resolution *after* the message
-        // which introduces it
-        Promise.resolve().then(() =>
-          resolveToKernel([[fresh, p.rejected, p.data]]),
-        );
-        cdebug(`comms export resolved ${lref} as ${fresh}`);
-        return fresh;
-      }
-
-      // Unresolved promises can use the same VPID until it is retired. Since
-      // we're telling the kernel about this VPID, we must arrange to notify
-      // the kernel when it resolves, unless the kernel itself is in control.
-      if (!deciderIsKernel(vpid)) {
-        subscribeKernelToPromise(vpid);
+    if (!state.toKernel.has(lref)) {
+      let kfref;
+      const { type } = parseLocalSlot(lref);
+
+      if (type === 'object') {
+        kfref = allocateKernelObjectID();
+      } else if (type === 'promise') {
+        const p = state.promiseTable.get(lref);
+        // We should always know about this lref. It is allocated upon
+        // receipt.  We retain the promiseTable entry even after the
+        // promise is resolved (to remember the resolution).
+        assert(p, X`how did I forget about ${lref}`);
+        kfref = allocateKernelPromiseID();
+        if (!p.resolved && !deciderIsKernel(lref)) {
+          subscribeKernelToPromise(lref);
+        }
+      } else {
+        assert.fail(X`unknown type ${type}`);
       }
-      return vpid;
+      state.toKernel.set(lref, kfref);
+      state.fromKernel.set(kfref, lref);
+      cdebug(`comms add mapping l->k ${kfref}<=>${lref}`);
     }
-    assert.fail(X`cannot give type ${type} to kernel`);
+    return state.toKernel.get(lref);
   }
 
-  function provideKernelForLocalResult(vpid) {
-    if (!vpid) {
+  function provideKernelForLocalResult(lpid) {
+    if (!lpid) {
       return null;
     }
-    const p = state.promiseTable.get(vpid);
-    assert(!p.resolved, X`result ${vpid} is already resolved`);
+    const p = state.promiseTable.get(lpid);
+    assert(!p.resolved, X`result ${lpid} is already resolved`);
     // TODO: reject somehow rather than crashing weirdly if we are not
     // already the decider, but I'm not sure how we could hit that case.
-    changeDeciderToKernel(vpid);
+    changeDeciderToKernel(lpid);
     // if we knew about this promise already, and thought the kernel
     // cared.. well, it doesn't now
-    unsubscribeKernelFromPromise(vpid);
-    return vpid;
+    unsubscribeKernelFromPromise(lpid);
+    return provideKernelForLocal(lpid);
+  }
+
+  function kernelSlotToLocalSlot(kernelSlot) {
```

Comment:
I was looking through the code to see if there was any place where this makes a difference (i.e., where failure for the mapping to already exist would be an error).  It looks to me like the translation of the message target in `sendToKernel` is such a case, so I will introduce the `get...` vs. `provide...` distinction for the kernel mapping as you describe.  This will also address the test case that I commented on above.